calyx-py/test/correctness/queues/binheap.py

```diff
@@ -29,107 +31,369 @@ def insert_swap(prog, name, width, len, idx_w):
+    def extract_fst(name, idx, out):
+        with comp.group(name) as extract_fst:
+            mem.addr0 = idx
+            untuplify.tup = mem.read_data
+            out.write_en = cb.HI
+            out.in_ = untuplify.fst
+            extract_fst.done = out.done
+
+        return extract_fst
+
+    # (_, out) := mem[indx]
+    def extract_snd(name, idx, out):
+        with comp.group(name) as extract_snd:
+            mem.addr0 = idx
+            untuplify.tup = mem.read_data
+            out.write_en = cb.HI
+            out.in_ = untuplify.snd
+            extract_snd.done = out.done
+
+        return extract_snd
+
+    extract_current_rank = extract_fst("extract_current_rank", 
+                                       current_idx.out, 
+                                       current_rank)
+    extract_parent_rank = extract_fst("extract_parent_rank", 
+                                      parent_idx.out, 
+                                      parent_rank)
+    extract_child_l_rank = extract_fst("extract_child_l_rank", 
+                                       child_l_idx.out, 
+                                       child_l_rank)
+    extract_child_r_rank = extract_fst("extract_child_r_rank", 
+                                      child_r_idx.out, 
+                                      child_r_rank)
+    
+    # current_rank < parent_rank
+    current_lt_parent = comp.lt_use(current_rank.out, parent_rank.out)
+    
+    le_1 = comp.le(factor)
+    le_2 = comp.le(32)
+    if_or = comp.or_(1)
+    with comp.comb_group("child_l_swap") as child_l_swap:
+        # Check if the `current_idx`th element should be swapped with its left child.
+        # size <= child_r_idx OR child_l_rank <= child_r_rank
+        le_1.left = size.out
+        le_1.right = child_r_idx.out
+
+        le_2.left = child_l_rank.out
+        le_2.right = child_r_rank.out
+
+        if_or.left = le_1.out
+        if_or.right = le_2.out
+
+    lt_1 = comp.lt(factor)
+    lt_2 = comp.lt(32)
+    lt_3 = comp.lt(factor)
+    lt_4 = comp.lt(32)
+    and_1 = comp.and_(1)
+    and_2 = comp.and_(1)
+    while_or = comp.or_(1)
+    with comp.comb_group("current_gt_children") as current_gt_children:
+        # Check if the `current_idx`th element should be swapped with its left OR right.
+        # child_l_idx < size AND child_l_rank < current_rank
+        # OR
+        # child_r_idx < size AND child_r_rank < current_rank
+        lt_1.left = child_l_idx.out
+        lt_1.right = size.out
+
+        lt_2.left = child_l_rank.out
+        lt_2.right = current_rank.out
+
+        lt_3.left = child_r_idx.out
+        lt_3.right = size.out
+
+        lt_4.left = child_r_rank.out
+        lt_4.right = current_rank.out
+
+        and_1.left = lt_1.out
+        and_1.right = lt_2.out
+
+        and_2.left = lt_3.out
+        and_2.right = lt_4.out
+
+        while_or.left = and_1.out
+        while_or.right = and_2.out
+
+    peak = extract_snd("peak", 0, ans)
+
+    pop = [
```

Comment:
For some reason I never thought to just create Python-level lists per command and then have the "menu" down below be much simpler. Very cool!